From f5d50a13b7c7725faf1c0a58b2c88d25398dd9de Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Wed, 17 Jul 2019 17:14:15 +0800
Subject: [PATCH] drm: add loader_protect for restore_fbdev_mode

the following flow is conflict with kernel logo display:
so we ignore the restore_fbdev_mode_atomic when in kernel
logo on state.

[    0.861904] vop_plane_atomic_update+0x2604/0x390c
[    0.861913] drm_atomic_helper_commit_planes+0x158/0x1d8
[    0.861922] rockchip_atomic_commit_complete+0x94/0xf0
[    0.861930] rockchip_drm_atomic_commit+0x1c0/0x1f8
[    0.861938] drm_atomic_commit+0x6c/0x84
[    0.861948] restore_fbdev_mode_atomic+0x1a8/0x224
[    0.861956] restore_fbdev_mode+0x60/0x168
[    0.861965] drm_fb_helper_restore_fbdev_mode_unlocked+0x74/0xb0
[    0.861974] drm_fb_helper_set_par+0x80/0xa0
[    0.861983] drm_fb_helper_hotplug_event+0x10c/0x124
[    0.861991] rockchip_drm_output_poll_changed+0x54/0x64
[    0.862000] drm_kms_helper_hotplug_event+0x48/0x70
[    0.862009] output_poll_execute+0x1c4/0x214
[    0.862018] process_one_work+0x208/0x3bc
[    0.862027] worker_thread+0x214/0x318
[    0.862035] kthread+0x120/0x128
[    0.862043] ret_from_fork+0x10/0x18

Change-Id: I8e9874659e9f6b3a29e34f247855b73faf42fde0
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/drm_fb_helper.c | 3 ++-
 include/drm/drm_mode_config.h   | 5 +++++
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/drm_fb_helper.c b/drivers/gpu/drm/drm_fb_helper.c
index 249985b99a72..cf5eaf17e956 100644
--- a/drivers/gpu/drm/drm_fb_helper.c
+++ b/drivers/gpu/drm/drm_fb_helper.c
@@ -448,7 +448,8 @@ static int restore_fbdev_mode_atomic(struct drm_fb_helper *fb_helper, bool activ
 		}
 	}
 
-	ret = drm_atomic_commit(state);
+	if (!dev->mode_config.loader_protect)
+		ret = drm_atomic_commit(state);
 
 out_state:
 	if (ret == -EDEADLK)
diff --git a/include/drm/drm_mode_config.h b/include/drm/drm_mode_config.h
index 5abf51a489a5..35d2bcdeff0e 100644
--- a/include/drm/drm_mode_config.h
+++ b/include/drm/drm_mode_config.h
@@ -860,6 +860,11 @@ struct drm_mode_config {
 	struct drm_atomic_state *suspend_state;
 
 	const struct drm_mode_config_helper_funcs *helper_private;
+	/**
+	 * @loader_protect
+	 * ignore restore_fbdev_mode_atomic when in logo on state
+	 */
+	bool loader_protect;
 };
 
 void drm_mode_config_init(struct drm_device *dev);
-- 
2.35.3

